exercises/hamming/test/Tests.hs

```diff
@@ -102,3 +110,65 @@ cases = [ Case { description = "empty strands"
                , expected    = Nothing
                }
         ]
+
+hammingGen :: Int -> Gen Case
+hammingGen 0 = return
+  Case { description = "Empty strands have no distance"
+       , strand1 = ""
+       , strand2 = ""
+       , expected = Just 0
+       }
+
+hammingGen n = do
+  (a, b, count) <- nucleotidePairGen
+  case' <- hammingGen (n-1)
+  return . hammingDesc $
+    case' { description = ""
+          , strand1 = a ++ strand1 case'
+          , strand2 = b ++ strand2 case'
+          , expected = (+) <$> count <*> expected case'
+          }
+
+hammingDesc :: Case -> Case
+hammingDesc case'@Case{..} = case' { description = desc' }
+  where
+    desc' = "Strands of " ++ length' ++ " with " ++ dist' ++ " distance"
+    dist' = maybe "invalid" show expected
+    length' = case comparing length strand1 strand2 of
+      EQ -> "lengths " ++ show (length strand1)
+      _  -> "different lengths " ++ show (length strand1) ++ " and " ++
+            show (length strand2)
+
+nucleotidePairGen :: Gen (String, String, Maybe Integer)
+nucleotidePairGen = frequency
+  [ (20, zeroDistance)
+  , (4, oneDistance)
+  , (1, invalidDistance)
+  ]
+  where
+    zeroDistance = do
+      a <- elements "ACGT"
+      return ([a], [a], Just 0)
+
+    oneDistance = do
+      a <- elements "ACGT"
+      b <- elements ("ACGT" \\ [a])
+      return ([a], [b], Just 1)
+
+    invalidDistance = do
+      a <- elements "ACGT"
+      return ([a], [], Nothing)
+
+without :: [a] -> [(a, [a])]
```

Comment:
note for me, in case I forget when coming back to it later:

`without [1, 2, 3] = [(1,[2,3]),(2,[1,3]),(3,[1,2])]`

Reply:
I'll add a comment and name it `select` from Prolog.